File: sdk/ios/Sources/Afftok/Afftok.swift
import Foundation

/// AffTok SDK - Main Entry Point
///
/// The AffTok SDK provides click and conversion tracking for iOS applications.
/// It supports offline queuing, automatic retry with exponential backoff,
/// HMAC-SHA256 signature verification, and zero-drop tracking.
///
/// Usage:
/// ```swift
/// // Initialize
/// Afftok.shared.initialize(options: AfftokOptions(
///     apiKey: "your_api_key",
///     advertiserId: "your_advertiser_id",
///     debug: true
/// ))
///
/// // Track click
/// await Afftok.shared.trackClick(params: ClickParams(offerId: "offer123"))
///
/// // Track conversion
/// await Afftok.shared.trackConversion(params: ConversionParams(
///     offerId: "offer123",
///     transactionId: "txn_abc123",
///     amount: 29.99
/// ))
/// ```
public class Afftok {
    
    /// Shared singleton instance
    public static let shared = Afftok()
    
    private var isInitialized = false
    private var options: AfftokOptions?
    private var queue: OfflineQueue?
    private var clickTracker: ClickTracker?
    private var conversionTracker: ConversionTracker?
    
    private init() {}
    
    /// Initialize the AffTok SDK
    ///
    /// - Parameter options: SDK configuration options
    public func initialize(options: AfftokOptions) {
        guard !isInitialized else {
            log("SDK already initialized")
            return
        }
        
        self.options = options
        
        // Initialize components
        queue = OfflineQueue(debug: options.debug)
        clickTracker = ClickTracker(options: options, queue: queue!)
        conversionTracker = ConversionTracker(options: options, queue: queue!)
        
        // Start auto-flush if enabled
        if options.autoFlush {
            startAutoFlush()
        }
        
        isInitialized = true
        log("SDK initialized successfully")
        log("Device ID: \(DeviceInfo.getDeviceId())")
        log("Pending queue items: \(queue?.size() ?? 0)")
    }
    
    /// Track a click event
    ///
    /// - Parameter params: Click parameters
    /// - Returns: AfftokResponse with tracking result
    public func trackClick(params: ClickParams) async -> AfftokResponse {
        guard let tracker = clickTracker else {
            return AfftokResponse(success: false, error: "SDK not initialized")
        }
        return await tracker.trackClick(params: params)
    }
    
    /// Track a click event with completion handler
    public func trackClick(params: ClickParams, completion: @escaping (AfftokResponse) -> Void) {
        Task {
            let response = await trackClick(params: params)
            DispatchQueue.main.async {
                completion(response)
            }
        }
    }
    
    /// Track a click with a signed tracking link
    ///
    /// - Parameters:
    ///   - signedLink: The signed tracking link from AffTok
    ///   - params: Additional click parameters
    /// - Returns: AfftokResponse with tracking result
    public func trackSignedClick(signedLink: String, params: ClickParams) async -> AfftokResponse {
        guard let tracker = clickTracker else {
            return AfftokResponse(success: false, error: "SDK not initialized")
        }
        return await tracker.trackSignedClick(signedLink: signedLink, params: params)
    }
    
    /// Track a conversion event
    ///
    /// - Parameter params: Conversion parameters
    /// - Returns: AfftokResponse with tracking result
    public func trackConversion(params: ConversionParams) async -> AfftokResponse {
        guard let tracker = conversionTracker else {
            return AfftokResponse(success: false, error: "SDK not initialized")
        }
        return await tracker.trackConversion(params: params)
    }
    
    /// Track a conversion event with completion handler
    public func trackConversion(params: ConversionParams, completion: @escaping (AfftokResponse) -> Void) {
        Task {
            let response = await trackConversion(params: params)
            DispatchQueue.main.async {
                completion(response)
            }
        }
    }
    
    /// Track conversion with additional metadata
    public func trackConversionWithMeta(params: ConversionParams, metadata: [String: Any]) async -> AfftokResponse {
        guard let tracker = conversionTracker else {
            return AfftokResponse(success: false, error: "SDK not initialized")
        }
        return await tracker.trackConversionWithMeta(params: params, metadata: metadata)
    }
    
    /// Manually enqueue an event for later processing
    ///
    /// - Parameters:
    ///   - type: Event type ("click" or "conversion")
    ///   - payload: Event payload
    /// - Returns: Queue item ID
    @discardableResult
    public func enqueue(type: String, payload: [String: Any]) -> String {
        guard let q = queue else { return "" }
        return q.enqueue(type: type, payload: payload)
    }
    
    /// Manually flush the offline queue
    public func flush() async {
        guard let q = queue else { return }
        await q.flush { [weak self] item in
            await self?.processQueueItem(item: item) ?? false
        }
    }
    
    /// Flush queue with completion handler
    public func flush(completion: @escaping () -> Void) {
        Task {
            await flush()
            DispatchQueue.main.async {
                completion()
            }
        }
    }
    
    /// Get the device fingerprint
    public func getFingerprint() -> String {
        return DeviceInfo.getFingerprint()
    }
    
    /// Get the device ID
    public func getDeviceId() -> String {
        return DeviceInfo.getDeviceId()
    }
    
    /// Get device info
    public func getDeviceInfo() -> [String: String] {
        return DeviceInfo.getDeviceInfo()
    }
    
    /// Get number of pending queue items
    public func getPendingCount() -> Int {
        return queue?.size() ?? 0
    }
    
    /// Check if SDK is initialized
    public func isReady() -> Bool {
        return isInitialized
    }
    
    /// Get SDK version
    public func getVersion() -> String {
        return AfftokConfig.sdkVersion
    }
    
    /// Clear all pending queue items
    public func clearQueue() {
        queue?.clear()
    }
    
    /// Shutdown the SDK
    public func shutdown() {
        queue?.stopAutoFlush()
        isInitialized = false
        log("SDK shutdown")
    }
    
    // MARK: - Private Methods
    
    private func startAutoFlush() {
        guard let q = queue, let opts = options else { return }
        q.startAutoFlush(interval: opts.flushInterval) { [weak self] item in
            await self?.processQueueItem(item: item) ?? false
        }
    }
    
    private func processQueueItem(item: QueueItem) async -> Bool {
        switch item.type {
        case "click":
            let params = ClickParams(
                offerId: item.payload["offer_id"] ?? "",
                trackingCode: item.payload["tracking_code"],
                subId1: item.payload["sub_id_1"],
                subId2: item.payload["sub_id_2"],
                subId3: item.payload["sub_id_3"]
            )
            let response = await trackClick(params: params)
            return response.success
            
        case "conversion":
            let params = ConversionParams(
                offerId: item.payload["offer_id"] ?? "",
                transactionId: item.payload["transaction_id"] ?? "",
                clickId: item.payload["click_id"],
                amount: Double(item.payload["amount"] ?? ""),
                currency: item.payload["currency"] ?? "USD",
                status: item.payload["status"] ?? "pending"
            )
            let response = await trackConversion(params: params)
            return response.success
            
        default:
            log("Unknown event type: \(item.type)")
            return false
        }
    }
    
    private func log(_ message: String) {
        if options?.debug == true {
            print("[AffTok SDK] \(message)")
        }
    }
}

File: sdk/android/afftok-sdk/src/main/java/com/afftok/Afftok.kt
package com.afftok

import android.content.Context
import kotlinx.coroutines.*

/**
 * AffTok SDK - Main Entry Point
 * 
 * The AffTok SDK provides click and conversion tracking for Android applications.
 * It supports offline queuing, automatic retry with exponential backoff,
 * HMAC-SHA256 signature verification, and zero-drop tracking.
 * 
 * Usage:
 * ```kotlin
 * // Initialize
 * Afftok.init(context, AfftokOptions(
 *     apiKey = "your_api_key",
 *     advertiserId = "your_advertiser_id",
 *     debug = true
 * ))
 * 
 * // Track click
 * Afftok.trackClick(ClickParams(offerId = "offer123"))
 * 
 * // Track conversion
 * Afftok.trackConversion(ConversionParams(
 *     offerId = "offer123",
 *     transactionId = "txn_abc123",
 *     amount = 29.99
 * ))
 * ```
 */
object Afftok {
    private var isInitialized = false
    private lateinit var options: AfftokOptions
    private lateinit var context: Context
    private lateinit var fingerprintProvider: FingerprintProvider
    private lateinit var queue: OfflineQueue
    private lateinit var clickTracker: ClickTracker
    private lateinit var conversionTracker: ConversionTracker
    
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    
    /**
     * Initialize the AffTok SDK
     * 
     * @param context Android application context
     * @param options SDK configuration options
     */
    fun init(context: Context, options: AfftokOptions) {
        if (isInitialized) {
            log("SDK already initialized")
            return
        }
        
        this.context = context.applicationContext
        this.options = options
        
        // Initialize components
        fingerprintProvider = FingerprintProvider(this.context)
        queue = OfflineQueue(this.context, options.debug)
        clickTracker = ClickTracker(options, fingerprintProvider, queue)
        conversionTracker = ConversionTracker(options, fingerprintProvider, queue)
        
        // Start auto-flush if enabled
        if (options.autoFlush) {
            startAutoFlush()
        }
        
        isInitialized = true
        log("SDK initialized successfully")
        log("Device ID: ${fingerprintProvider.getDeviceId()}")
        log("Pending queue items: ${queue.size()}")
    }
    
    /**
     * Track a click event
     * 
     * @param params Click parameters
     * @return AfftokResponse with tracking result
     */
    suspend fun trackClick(params: ClickParams): AfftokResponse {
        ensureInitialized()
        return clickTracker.trackClick(params)
    }
    
    /**
     * Track a click event (non-suspend version using callback)
     */
    fun trackClick(params: ClickParams, callback: (AfftokResponse) -> Unit) {
        ensureInitialized()
        scope.launch {
            val response = clickTracker.trackClick(params)
            withContext(Dispatchers.Main) {
                callback(response)
            }
        }
    }
    
    /**
     * Track a click with a signed tracking link
     * 
     * @param signedLink The signed tracking link from AffTok
     * @param params Additional click parameters
     * @return AfftokResponse with tracking result
     */
    suspend fun trackSignedClick(signedLink: String, params: ClickParams): AfftokResponse {
        ensureInitialized()
        return clickTracker.trackSignedClick(signedLink, params)
    }
    
    /**
     * Track a conversion event
     * 
     * @param params Conversion parameters
     * @return AfftokResponse with tracking result
     */
    suspend fun trackConversion(params: ConversionParams): AfftokResponse {
        ensureInitialized()
        return conversionTracker.trackConversion(params)
    }
    
    /**
     * Track a conversion event (non-suspend version using callback)
     */
    fun trackConversion(params: ConversionParams, callback: (AfftokResponse) -> Unit) {
        ensureInitialized()
        scope.launch {
            val response = conversionTracker.trackConversion(params)
            withContext(Dispatchers.Main) {
                callback(response)
            }
        }
    }
    
    /**
     * Track conversion with additional metadata
     */
    suspend fun trackConversionWithMeta(
        params: ConversionParams,
        metadata: Map<String, Any>
    ): AfftokResponse {
        ensureInitialized()
        return conversionTracker.trackConversionWithMeta(params, metadata)
    }
    
    /**
     * Manually enqueue an event for later processing
     * 
     * @param type Event type ("click" or "conversion")
     * @param payload Event payload
     * @return Queue item ID
     */
    fun enqueue(type: String, payload: Map<String, Any>): String {
        ensureInitialized()
        return queue.enqueue(type, payload)
    }
    
    /**
     * Manually flush the offline queue
     */
    suspend fun flush() {
        ensureInitialized()
        queue.flush { item ->
            processQueueItem(item)
        }
    }
    
    /**
     * Flush queue with callback
     */
    fun flush(callback: () -> Unit) {
        ensureInitialized()
        scope.launch {
            flush()
            withContext(Dispatchers.Main) {
                callback()
            }
        }
    }
    
    /**
     * Get the device fingerprint
     */
    fun getFingerprint(): String {
        ensureInitialized()
        return fingerprintProvider.getFingerprint()
    }
    
    /**
     * Get the device ID
     */
    fun getDeviceId(): String {
        ensureInitialized()
        return fingerprintProvider.getDeviceId()
    }
    
    /**
     * Get device info
     */
    fun getDeviceInfo(): Map<String, String> {
        ensureInitialized()
        return fingerprintProvider.getDeviceInfo()
    }
    
    /**
     * Get number of pending queue items
     */
    fun getPendingCount(): Int {
        ensureInitialized()
        return queue.size()
    }
    
    /**
     * Check if SDK is initialized
     */
    fun isReady(): Boolean = isInitialized
    
    /**
     * Get SDK version
     */
    fun getVersion(): String = Config.SDK_VERSION
    
    /**
     * Clear all pending queue items
     */
    fun clearQueue() {
        ensureInitialized()
        queue.clear()
    }
    
    /**
     * Shutdown the SDK
     */
    fun shutdown() {
        if (!isInitialized) return
        
        queue.stopAutoFlush()
        scope.cancel()
        isInitialized = false
        log("SDK shutdown")
    }
    
    // Private methods
    
    private fun startAutoFlush() {
        queue.startAutoFlush(options.flushInterval) { item ->
            processQueueItem(item)
        }
    }
    
    private suspend fun processQueueItem(item: QueueItem): Boolean {
        return try {
            val response = when (item.type) {
                "click" -> {
                    val params = ClickParams(
                        offerId = item.payload["offer_id"] as? String ?: "",
                        trackingCode = item.payload["tracking_code"] as? String,
                        subId1 = item.payload["sub_id_1"] as? String,
                        subId2 = item.payload["sub_id_2"] as? String,
                        subId3 = item.payload["sub_id_3"] as? String
                    )
                    clickTracker.trackClick(params)
                }
                "conversion" -> {
                    val params = ConversionParams(
                        offerId = item.payload["offer_id"] as? String ?: "",
                        clickId = item.payload["click_id"] as? String,
                        transactionId = item.payload["transaction_id"] as? String ?: "",
                        amount = item.payload["amount"] as? Double,
                        currency = item.payload["currency"] as? String ?: "USD",
                        status = item.payload["status"] as? String ?: "pending"
                    )
                    conversionTracker.trackConversion(params)
                }
                else -> AfftokResponse(success = false, error = "Unknown event type")
            }
            response.success
        } catch (e: Exception) {
            log("Error processing queue item: ${e.message}")
            false
        }
    }
    
    private fun ensureInitialized() {
        if (!isInitialized) {
            throw IllegalStateException("AffTok SDK not initialized. Call Afftok.init() first.")
        }
    }
    
    private fun log(message: String) {
        if (::options.isInitialized && options.debug) {
            println("[AffTok SDK] $message")
        }
    }
}

